packages/x-data-grid-pro/src/hooks/features/dataSource/useGridDataSourceBasePro.ts

```diff
@@ -304,6 +317,56 @@ export const useGridDataSourceBasePro = <Api extends GridPrivateApiPro>(
     [apiRef],
   );
 
+  const handleGroupedDataUpdate = React.useCallback<GridStrategyProcessor<'dataSourceRowsUpdate'>>(
+    (params) => {
+      if ('error' in params) {
+        apiRef.current.setRows([]);
+        return;
+      }
+
+      const {
+        response,
+        options: { keepChildrenExpanded },
+      } = params;
+      if (response.rowCount !== undefined) {
+        apiRef.current.setRowCount(response.rowCount);
+      }
+
+      if (keepChildrenExpanded === false) {
+        apiRef.current.setRows(response.rows);
+      } else {
+        const tree = gridRowTreeSelector(apiRef);
+        // Remove existing outdated rows before setting the new ones
+        // Create a set of the current root rows
+        const parentRowsToDelete = new Set(
+          getTreeNodeDescendants(tree, GRID_ROOT_GROUP_ID, false, true),
+        );
+        // Remove from the list the rows that are again in the response
+        response.rows.forEach((row) => {
+          parentRowsToDelete.delete(gridRowIdSelector(apiRef, row));
+        });
+        const rowsToDelete: { id: GridRowId; _action: 'delete' }[] = [];
+        if (parentRowsToDelete.size > 0) {
+          parentRowsToDelete.forEach((parentRowId) => {
+            const descendants = getTreeNodeDescendants(tree, parentRowId, false, false).reverse(); // reverse to make sure that the deepest descendants are deleted first
+            for (let i = 0; i < descendants.length; i += 1) {
+              rowsToDelete.push({ id: descendants[i], _action: 'delete' });
+            }
+            rowsToDelete.push({ id: parentRowId, _action: 'delete' });
+          });
+        }
+        apiRef.current.updateRows([...rowsToDelete, ...response.rows]);
```

Comment:
If we anticipate arrays with hundreds of elements here, we'd be better off using `.concat` instead – see https://github.com/mui/mui-x/pull/19575#discussion_r2394919485